server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java

```diff
@@ -71,11 +72,11 @@ public static void prepareNewNamespaceState(ZooReaderWriter zoo, final PropStore
       InstanceId instanceId, NamespaceId namespaceId, String namespace,
       NodeExistsPolicy existsPolicy) throws KeeperException, InterruptedException {
     log.debug("Creating ZooKeeper entries for new namespace {} (ID: {})", namespace, namespaceId);
-    String zPath = Constants.ZROOT + "/" + instanceId + Constants.ZNAMESPACES + "/" + namespaceId;
+    String zPath = Constants.ZROOT + "/" + instanceId + Constants.ZNAMESPACES;
 
-    zoo.putPersistentData(zPath, new byte[0], existsPolicy);
-    zoo.putPersistentData(zPath + Constants.ZNAMESPACE_NAME, namespace.getBytes(UTF_8),
-        existsPolicy);
+    zoo.putPersistentData(zPath + "/" + namespaceId, new byte[0], existsPolicy);
+    zoo.putPersistentData(zPath,
+        NamespaceMapping.writeNamespaceToMap(zoo, zPath, namespaceId, namespace), existsPolicy);
```

Comment:
This code does not seems like it is handling concurrent updates to the new namespace node correctly. Can use use the mutateExisting function to do this.  This function will attempt to update a zk node and automatically retry if there was a change to the node between reading the data and writing the updated.  Also not sure why one of the writes to zookeeper is being done.

```suggestion
    // Why is the following be added to zookeeper?
    zoo.putPersistentData(zPath + "/" + namespaceId, new byte[0], existsPolicy);
    
    zoo.mutateExisting(zPath, data->{
      var namespaces = NamespaceMapping.deserialize(data);
      // TODO throw exception if namespace name already exists in map?
      namespaces.put(namespaceId.canonical(),namespace);
      return NamespaceMapping.serialize(namespaces);
    });
```

Reply:
@keith-turner I think this method is only ever called during the init code, which is in a single thread, and in the fate operation for creating a table, where there is a lock held. So, concurrency is already handled. That said, it probably couldn't hurt to use mutateExisting here, to make it safe for future refactorings of Fate.

Reply:
> @keith-turner I think this method is only ever called during the init code, which is in a single thread

It is also called by PopulateZookeeperWithNamespace which is part of the create table code and could call the code concurrently.

Reply:
Updated to use mutateExisting, included TODO to check for already-existing namespace name in future commit (likely part of this PR still).